### ISSUE TITLE
Internal error: cannot type-check file

### ISSUE THREAD

**Issue by nickcollins** [agda/agda#4833]
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Monad/Signature.hs:665

Agda version 2.6.1

I can attach a tarball of the folder if requested

**Comment by nickcollins** [comment]
I've managed to massage agda into not hitting this issue, and haven't seen it since. maybe had something to do with an out of sync agdai file after i updated agda? anyways just saying that this isn't an urgent or blocking bug for me, it may be closed if it's not of interest to anyone else

**Comment by nickcollins** [comment]
the error reappeared - it happens when I use a multiline comment to comment out everything after a certain point in a file. So still not an urgent problem, but maybe that helps shed light on what's going on

**Comment by nickcollins** [comment]
Actually, i hit the error again, under fairly ordinary circumstances (i.e. without commenting out a bunch of stuff), so now it actually is pretty urgent for me.

Note that when i comment out the `abstract` keyword that encompasses everything in my module, this problem, and another one i was having where it says some constructor is out of scope (it isn't), go away if i remove the use of `abstract`

**Comment by rwe** [comment]
For reference, the error is on line 664 here (linking to the `2.6.1` tag, though GitHub doesn't seem to inline previews for a non-hash ref): https://github.com/agda/agda/blob/8eb0d01811a663cf2b27b482b3b18690adfa094b/src/full/Agda/TypeChecking/Monad/Signature.hs#L659-L666

There are a few open issues with the [`abstract` label](https://github.com/agda/agda/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+abstract+label%3Aabstract). Sounds like this is possibly same as #3744 (_Internal error related to abstract_)? cc @andreasabel @nad @UlfNorell

**Comment by andreasabel** [comment]
Possibly.
A MWE demonstrating the bug would be helpful to clarify... @nickcollins, could you supply one?

**Comment by nickcollins** [comment]
I hit this bug during a transitional point in developing the code, so now with my code complete (save for the `abstract` keyword being commented out) I don't hit this bug at all, let alone have an MWE.

However, if i uncomment `abstract`, I do hit a different bug: 
```
/home/nick/dependent-dicts/agda/Delta.agda:116,5-651,36
Not in scope:
  Delta._∈'_.InH
  at /home/nick/dependent-dicts/agda/Delta.agda:261,53-56
    (did you mean
       '_∨_.Inl' or
       ... many other suggestions
       'InH' or
       'InT' or
       'Inl' or
       'Inr'?)
```

For this variety of bug, i have an MWE that reprod in 2.6.0.1, but was fixed in 2.6.1:
```
module m where
   abstract
     data d : Set where
       foo : d
       bar : d
 
     f1 : d → d
     f1 x with x
     ... | foo = bar -- the use of either foo or bar as the result here causes the "Not in scope" error in 2.6.0.1
     ... | bar = ? -- Filling this hole with either foo or bar will work just fine, but then type checking will fail
 
     f2 : d → d
     f2 _ = foo -- note there's no bug here because it's not under a 'with'
```
It appears to me that one flavor of this bug was fixed in 2.6.1, but that other flavors, like the one my current code hits, still persist. I can't feasibly cut an MWE from this bug because my code, and the apparent circumstances of its appearance, are too complicated. But I suspect that reference to the 2.6.0.1 bug, and its solution, may lead to figuring out the current problem.

These two bugs are distinct, as one causes a fatal and the other causes an invalid "Not in scope" complaint. But I believe they're related, because at the time I hit the bug I initially reported, i could toggle a slight change in my code and instead get the "Not in scope" bug

For reference, my current code is at https://github.com/nickcollins/dependent-dicts-agda/tree/f9324f2acb39517ef15d27026b8be8f22201105f

**Comment by uncle-betty** [comment]
I think that I might have an MWE. The trick seems to be to tickle the termination checker, which the above MWE doesn't do. Here's what I use:

```
module Repro where
  abstract
    data Nat : Set where
      Zero : Nat
      Succ : Nat → Nat

    countDown : Nat → Nat
    countDown x with x
    ... | Zero   = Zero
    ... | Succ n = countDown n
```

The important part seems to be the recursion after the `with`. Type-checking with 2.6.1 or with current master yields:

```
Checking Repro (/home/tlo/dependent-dicts-agda/Repro.agda).
/home/tlo/dependent-dicts-agda/Repro.agda:7,5-10,31
Not in scope:
  Repro.Nat.Zero at /home/tlo/dependent-dicts-agda/Repro.agda:9,11-15
    (did you mean 'Zero'?)
```

Commenting out the `abstract` in the above MWE makes the issue go away.

Looking at the data structures, I find the definition of `countDown` to have `funAbstr = AbstractDef`, as expected. In contrast, the generated helper for `with` unexpectedly has `funAbstr = ConcreteDef`. So, when `termDef` termination-checks the `with` helper, `inConcreteOrAbstractMode` makes the termination check run in concrete mode, which precludes access to the data constructors.

Unfortunately, I've now reached the limits of my understanding of the Agda code. I'll have to leave it to the grown-ups to devise a fix.
